### DISCUSSION TITLE
How does a receiver Actor know who is the sender Actor

### DISCUSSION THREAD

**Discussion by nsengupta** [slawlor/ractor#127] [Q&A] [answered]
I am just starting with `ractor`. I have reasonably well background of Akka ( I had built production-systems using it ) but almost nothing of Erlang's Actor model.

My intention is to write a few articles as tutorials for `ractor` and I hope that in the process, I will have deeper understanding of how is it supposed to be used. 

If an Actor receives a message - say 'Ping' - and wants to respond to the sender with a 'Pong' - how does it know who the sender is? In Akka, the `sender` ActorRef is automatically available (the receiver actor may or may not use it, but it is available). What will be an equivalent structure / operation (or both) to accomplish this _pattern_?

Pardon me, if this is too much of a beginner's question.

**Comment by nsengupta** [comment]
After going through 
* this section of README ( '' Deriving the basic Message trait for in-process only actors "),
* a few beginner's document on Erlang, and
* Ractor playground

My understanding is that the protocol for intra-actor conversation (on the same, local Actor runtime) will have to **fold in** the sender's `ActorRef` in the message itself. That way, the receving actor has no problem in knowing who should it send a response to.

Not sure, if I am going in the right direction though.

**Reply by slawlor** [reply]
Sorry for the delay, I'm on a holiday. But you are correct, by default the receiver has no notion of the sender. In that way you would need to send the actor ref to the receiver so that they could then know who made the call. If you only want to identify the actor, you could send just the actor ID but if you want to use the ref to send messages in the future, you do have to fold it in

**Reply by nsengupta** [reply, marked as answer]
Many thanks, @slawlor m for the affirmation. Don't worry about the delay. 

I have begun writing, what I consider to be a tutorial for `ractor` and my objective is to _ease_ someone interested about Actors, _into_ `ractor`.

**Reply by slawlor** [reply]
Hello again @nsengupta I finally had a change to read your tutorial, and I thought it's really well done! I was wondering if you'd like to put up a PR against our `gh-pages` branch in order to publish it on our github site :) I think it's helpful information for newcomers to actors and `ractor` in Rust!

Thanks again for the great tutorial :)
